Makefile.local

```diff
@@ -2,23 +2,85 @@
 # specific to github.com/projectcalico/api. This is opposed to Makefile, which
 # is mirrored from github.com/projectcalico/calico/api.
 
+SHELL = /bin/bash -eu
+
+# These are local variables we use.
+LOCAL_REPOSITORY_REVISION ?= $(shell git describe --tags)
+GIT_BRANCH := $(shell git branch --show-current)
+PR_BRANCH := auto_api_update_$(shell git branch --show-current | sed -E -e 's/release-(calient)?-//')
+ORGANIZATION := tigera
+CALICO_REPOSITORY_NAME := tigera/calico-private
+
+CALICO_TEMP_DIR := /tmp/calico-api-mirror
+
+KEEP_LOCAL_FILES := README.md
+
+VERBOSE_LOCAL_VARIABLES = LOCAL_REPOSITORY_REVISION GIT_BRANCH PR_BRANCH CALICO_TEMP_DIR CALICO_REPOSITORY_NAME
+
+help:
+	$(info Calico API local Makefile)
+	$(info )
+	$(info Targets:)
+	$(info * variables:   Print the current, most relevant local variables)
+	$(info * update:      Clone upstream calico and bring their API files in)
+	$(info * pr:          Pulls changes from upstream calico, commits them, and creates a new PR for them.)
+	@true
+
+variables:
+	$(foreach localvar,$(VERBOSE_LOCAL_VARIABLES),$(info $(localvar) = $($(localvar))))
+	@true
+
+pr: update restore-local-files commit-remote-api show-pr-url
```

Comment:
`restore-local-files` is a dependency of `commit-remote-api` and is also called by `update`, so I don't think we need it here. It's also a small flag that perhaps the decomposition of the `update` target isn't quite right. I have a suspicion we might be breaking it down too much. When are we going to want to call these things independently of each other? Is the extra complexity required to track which targets depend on each other worth it in this case?

I tend to like this type of split for "real" make targets that produce an output in order to reduce duplication of logic, but all of these are PHONY targets that don't actually generate an output, so treating them as dependencies of each other doesn't seem well suited to Makefile and defeats the main purpose of the design pattern of using smaller targets that compose into something else.

Reply:
The complicated thing here is that nothing in this Makefile produces a singular 'target' that we can point to, at least not that I can see.

The `update` target (as it was and as it is in this PR) pulls in new copies of files to replace the existing copies, but the repository doesn't have any source files that it generates from; that means `make` can't do any calculation on whether or not anything needs done, so the target sort of has to be `PHONY` by definition.

The `pr` target does create an output (the PR) but not one that `make` is able to track, meaning it has to be a `PHONY` target as well. We could theoretically create arbitrary stamp files (e.g. `.pr_created_$(VERSION).stamp`) which depend on the API files, but that seems even messier and more complicated.

From my view, this Makefile is really just a shell script with a wrong interpreter. It doesn't make any decisions about what needs doing, doesn't track inputs or outputs, and doesn't really need to do dependency calculation either. It's really just "run this shell code" (update the API files) or "run that code and also this code too" (update API files and create the PR). 

In that context, I've broken things up into two targets; one to update the API files (for cases where you just want to see if there are changes) and one to create a PR (which I think inherently requires updating the API files also). I think in the context of what this Makefile is doing (running blocks of shell code), this makes sense.

I further broke things up into smaller, simpler targets (e.g. `show-pr-url`) partly for easier debugging for myself, but also to make the Makefile more readable. I realize that using `make` targets as functions is non-idiomatic and not the intended use of the tool (though it could be argued that the existence of `PHONY` targets disputes this), but in the context of "this is the tool we're using" it feels like the best option to me.

The only other approach that would make sense to me would be to move all of this shell code into a shell script with proper functions, etc. and just have `Makefile.local` call that script depending on what the user is trying to do (or remove `Makefile.local` entirely).

I'm open to suggestions on how things are structured overall, but given the choice between re-composing tasks into one or two large targets or keeping them separate for readability, I'd generally prefer the latter.

Reply:
> From my view, this Makefile is really just a shell script with a wrong interpreter.

Maybe it really should just be this, then :grin: 

```
update:
  ./perform-update.sh
```

But my point wasn't that PHONY targets are invalid in some way, just that the way the decomposition was structured was wrong - it was resulting in the wrong behavior, because PHONY targets that need only be called once were being called multiple times, which made actually _decreased readability_ substantially. 

I think it's fine to keep these targets PHONY and split up, but they shouldn't be used as dependencies of multiple other targets that are then dependencies of each other.